poly-org.el

```diff
@@ -42,7 +42,18 @@
 (define-obsolete-variable-alias 'pm-inner/org 'poly-org-innermode "v0.2")
 
 (defun poly-org-mode-matcher ()
-  (when (re-search-forward "#\\+begin_src +\\([^ \t\n]+\\)" (point-at-eol) t)
+  (when (re-search-forward
+         ;; there is no exhaustive list of block names in org-mode source
+         ;; the most complete list would be
+         ;; the default value of org-structure-template-alist
+         ;; we took everything listed there into account
```

Comment:
Maybe there is a more direct way to call the org-mode search engine instead of relying on the regexp?

Reply:
There is `org-next-block`. It returns point at beginning of the opening line of found block and throws an error if no block is found. `poly-org-mode-matcher`, on the other hand, sets point to the end of the occurrence found and explicitly never throws errors.

`org-next-block` will by default find e.g. dynamic blocks too. This is likely undesirable here. `org-next-block` provides a way to customize its search but overall it looks like using `org-next-block`, while possible in principle, won't help much. It will certainly make the code more verbose.

It is also not clear to me what `poly-org-mode-matcher` actually does. Its name suggests that it matches but why does it perform side-effects then? Matching is pure. Also, I don't have a good mental model of state changes in Emacs Lisp when it comes to `(point)` and friends. Thus, polishing the patch to ensure quality better be done by somebody else. On a purely procedural level, it works.

Reply:
What "side-effects" do you see? 

 It's used as a function in `:mode-matcher` slot. All those slots are documented in the [`polymode-classes.el`](https://github.com/polymode/polymode/blob/9eb9dce9c9a1d8c92e837818f576463c5bcf8952/polymode-classes.el#L30)

The idea is that once the head is matched `mode-matcher` retrieves the emacs mode name to install in that auto-mode. For auto-inner-modes the mode is not known in advance.

Reply:
> What "side-effects" do you see?

`poly-org-matcher` calls `re-search-forward` which, according to its documentation, “sets point”.

> It's used as a function in :mode-matcher slot

I see. I didn't know about `:documentation` for slots, cool.

Dacumentation says “Function is called at the beginning of the head span”. It's inappro4riate to call `org-next-block` here, as it searches from arbitrary point. I just encountered lots of `org-element-<block-name>-block-parser` functions, with duplicated code. :-( `org-element-src-block-parser` will return, among other thitgs, the language string in an alist. But there is no generic (“generic”) function for arbitrary blocks which is what my patch, at least, implicitly needs. I'll dig further.

I guess there has to be org function that actually returns the relevant /mode/ since org knows how to font-lock, and it's the mode, not the string, that polymode needs in the end. That would be a very direct solution but it bypasses the explicit “get string” part and thus might lead to significant redesign of poly-org. I could try to implement it unless you believe it's a bad idea.

Reply:
>  It's inappropriate to call org-next-block here, as it searches from arbitrary point.

Let's not do anything fancy here. I think supporting src, examples and export is good enough and regexp based solution is decent. I have just pushed that. Thanks for bringing this up!